packages/public-api/src/website/service.ts

```diff
@@ -1,80 +1,120 @@
-    await webpage.save();
+  let hostnameEntity = await hostnameRepo.findOne({
+    hostname: url.hostname,
+  });
+  if (!hostnameEntity) {
+    hostnameEntity = await hostnameRepo.save({
+      hostname: url.hostname,
+    });
   }
 
-  // Save packages if ready
-  if (webpage.status === WebPage.Status.Processed && internal.detectedPackages.length > 0) {
-    await getRepository(WebPagePackage).upsert(
-      internal.detectedPackages.map((pkg) => ({
-        latestUrl: internal.url,
-        hostname: getHostnameFromUrl(internal.url),
-        packageName: pkg.name,
-        possiblePackageVersions: pkg.possibleVersions,
-        packageVersionRange: pkg.versionRange,
-        packageMetadata: {
-          approximateByteSize: pkg.approximateSize ?? undefined,
-        },
-      })),
-      ['hostname', 'packageName']
-    );
+  let webPageEntity = await webPageRepo
+    .createQueryBuilder('web_page')
+    .where('web_page.hostname_id = :hostnameId', { hostnameId: hostnameEntity.id })
+    .andWhere('web_page.path = :path', { path: url.pathname })
+    .limit(1)
+    .getOne();
+  if (!webPageEntity) {
+    webPageEntity = await webPageRepo.save({
+      hostname: hostnameEntity,
+      path: url.pathname,
+    });
   }
+
+  return webPageEntity;
 }
 
-export async function getWebPagesByHostname(hostname: string) {
-  return getRepository(WebPage).find({ hostname });
+export async function getOrRequestWebPageScan(url: string) {
+  const parsedUrl = new URL(url);
+
+  const db = await getDatabaseConnection();
+
+  const result = await db.transaction(async (em) => {
+    const webPageScanRepo = em.getRepository(WebPageScan);
+
+    const webPageEntity = await findOrCreateWebPage(parsedUrl, em);
+
+    const mostRecentScan = await webPageScanRepo
+      .createQueryBuilder('scan')
+      .where('scan.web_page_id = :webPageId', { webPageId: webPageEntity.id })
+      .orderBy('scan.created_at', 'DESC')
+      .limit(1)
+      .getOne();
+
+    if (mostRecentScan && Date.now() - mostRecentScan.createdAt.getTime() < RESCAN_TIMEOUT_MS) {
+      return mostRecentScan;
+    }
+
+    const webPageScanEntity = await webPageScanRepo.save({
+      webPage: webPageEntity,
+      status: WebPageScan.Status.Pending,
+    });
+
+    await internalApi.requestWebPageScan(parsedUrl.toString(), webPageScanEntity.id.toString());
+
+    return webPageScanEntity;
+  });
+
+  return result;
 }
 
-export async function getPackagesByHostname(hostname: string) {
-  return getRepository(WebPagePackage).find({
-    relations: ['registryMetadata'],
-    where: { hostname },
+export async function syncWebPageScanResult(scanReport: SystemApi.ScanReport) {
+  const db = await getDatabaseConnection();
+
+  return await db.transaction(async (em) => {
+    const webPageScanRepo = em.getRepository(WebPageScan);
+
+    let scanEntity: WebPageScan;
+    if (scanReport.id) {
+      scanEntity = await webPageScanRepo.findOneOrFail({ id: parseInt(scanReport.id, 10) });
```

Comment:
Can you please re-check typeorm type for `bigint`? From my experience it is always represented as a `string`.